### ISSUE TITLE
Provide API access to get and update organisation logo images

### ISSUE THREAD

**Issue by chrisinmtown** [MISP/MISP#7066]
Organisation logos today are stored as files below /var/www/MISP/app/webroot/img/orgs. A fetch via the API/PyMISP does not get the image bytes, image URL or even an indicator that a logo is defined.  I cannot find a way to push an org logo via the API/PyMISP.

**Describe the solution you'd like**

* Extend server and PyMISP to allow fetch, update of logos as base64 encoded strings
* Extend server to allow fetch of logo as binary string for service to web UI
* Show logos on the web UI by using endpoint described above

And possibly someday:

* Store logos in a blob column in the organisations table.

Because splitting user data across multiple stores means backups or transfers are complex.

**Comment by legoguy1000** [comment]
Wouldn't it just be easier and more efficient to update the org model to provide a URL for the org logo if it exists and then pymisp can be extended to download it if wanted? Storing blobs in a DB is doable but also not the most efficient.

**Comment by iglocska** [comment]
Yeah - even easier we can abstract it and load / save b64 encoded logos in the org object automatically. The storage shouldn't have any impact on that.

**Comment by chrisinmtown** [comment]
@iglocska I am not sure I follow you here, "abstract it", "storage shouldn't have any impact", sorry.

@legoguy1000 I like your suggestion to extend the server API to include the server's URL of the organisation logo within the get-org JSON response instead of a long base64-encoded field.  Probably PyMISP clients rarely need the logo bytes. But that change will make the get/add/update methods asymmetric for that field. In other words from PyMISP you will get a URL for the logo but you cannot push a URL for the logo or in any way update that field. I believe the server will need a separate endpoint to accept an org logo's bytes, right?

**Comment by iglocska** [comment]
From an API perspective you don't need to worry about the storage engine. The API acts as an abstraction layer, we can just add a new logo field on the organisation object which expects/includes the logo image - which we can store in whatever way we see fit behind the scenes, hence having no impact on how pymisp/your scripts would interact with the data.

**Comment by chrisinmtown** [comment]
@iglocska thanks for explaining, I think I understand now.  I'll update the title to reflect the underlying requirement: API access to get and update the organisation logo.

**Comment by legoguy1000** [comment]
> @iglocska I am not sure I follow you here, "abstract it", "storage shouldn't have any impact", sorry.
> 
> @legoguy1000 I like your suggestion to extend the server API to include the server's URL of the organisation logo within the get-org JSON response instead of a long base64-encoded field. Probably PyMISP clients rarely need the logo bytes. But that change will make the get/add/update methods asymmetric for that field. In other words from PyMISP you will get a URL for the logo but you cannot push a URL for the logo or in any way update that field. I believe the server will need a separate endpoint to accept an org logo's bytes, right?

APIs are usually asymmetric.  There is usually a lot of fields that only exist after an object is created or is dynamically generated from other data and not able to be modified by the user.

**Comment by chrisinmtown** [comment]
I want to make it clear I'm still in favor of storing *all* data in the database, not the filesystem, including images in blob-type columns. Why?  We use a primary-standby pair of servers for the MISP webapp.  Because of the project's odd choice to store MOST data in the database yet keep a LITTLE BIT of data on the local filesystem, we have to replicate a few directories from primary server to standby server. If the timing of a change is just right, of course that change can be lost.  It would be a lot cleaner & easier to implement a highly reliable system if the MISP webapp doesn't drop user-uploaded data into its filesystem.

**Comment by legoguy1000** [comment]
I would say u would definitely want to restrict the size of images that could be uploaded then to prevent it exploding the size of the DB.

**Comment by chrisinmtown** [comment]
Well sure @legoguy1000 but adding/updating organisations is restricted to admin role.

**Comment by JoePJisc** [comment]
Just stumbled upon this while trying to write a Python script to take a clean install to our production build, ready for a rebuild of our MISP instances. 

Being able to add logos via PyMISP (in whatever form), would be really usful for this.
